### ISSUE TITLE
Training Default Transformer LM  fails when using customized corpus

### ISSUE THREAD

**Issue by xiabingquan** [espnet/espnet#4508]
### Problem Description
> I'll first describe my problems in this section and provide more details in the next sections.

Recently I've been training a **char-level** language model for my own experiment. It's highly customized, so I changed the codes a little bit and only train the language model, but unfortunately that the training fails for some reasons. There're three problems I encountered:
1. About the loss. The cross entropy Loss declines super quickly, then plataeus, and then becomes NaN. There're about 20k iterations per epoch. The cross entropy loss declines from 3.6 to 1.3 in the first 10k iterations and then plataeus at about 1.1~1.3. In the about 13k iterations of the 2nd epoch, the loss NaNs, then the log constantly reports `WARNING: The grad norm is nan. Skipping updating the model` till the end of training. The loss trend from tensorboard is as followes
![image](https://user-images.githubusercontent.com/53991699/178402852-9359b6ac-9b26-4d52-ae46-70f7bcae9a66.png)
Note that the last 1/3 are all NaNs, 
2. About the perplexity. The log reports that the perplexity of test set is 3.2165575309226586: `2022-07-11 21:15:34,169 (lm_calc_perplexity:130) INFO: PPL=3.2165575309226586`. I think it's barely wrong, but I'm not sure what the problem is since I didn't change the codes of computing perplexity.
4. About the batch_bins, I used the default configuration file [conf/tuning/train_lm_transformer2](https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_lm_transformer2.yaml). The comments says that "This configuration requires Tesla V100-SXM2(32GB) x 16 GPUs It takes about 2 days.". In my end, I use2 GeForce RTX™ 3090(24GB) × 2 GPUS. The total gpu memory is about 24*2/32/16~=0.1 of the V100 setting. However, the maximum of  `batch_bins ` for me is only 48k (otherwise OOM occurs), far less than 500000k * 0.1 = 50000k. I'm wondering why I couldn't set a large `batch_bins`.

### Details

The code changes I made are as follows
- Modify [asr.sh](https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/asr.sh). Remove unrelated parts and only reserve the codes of training language models
- Modify some codes according to my customized needs, such as changing the `sos`, `eos`, `ignore_id` and so on. I'm pretty sure I've changed all relavant settings about those special tokens(such as token padding, loss calculation). The default ESPnet uses `<sos/eos>` to represent both `<sos>` and `<eos>` while I use `<sos>` and `<eos>` respectively, but I don't think this change will make a difference.
- Use my own corpus. It consists of the transcriptions of [LRS2](https://www.robots.ox.ac.uk/~vgg/data/lip_reading/lrs2.html), [LRS3](https://www.robots.ox.ac.uk/~vgg/data/lip_reading/lrs3.html) and [LibriSpeech](https://www.openslr.org/12). The whole corpus contains 534335 sentences. I randomly split the whole corpus into training, development, and test set. The development and test set both contains 1000 sentences, and the training set contains the rest, for a total of 532337 sentences.

The lines of the training corpus look like
> 0 WHAT IS IT AND HOW OLD IS IT
1 SOON MADE HER CONFESS THAT SHE HAD NEVER SEEN HIM THEN THEY PROCEEDED TO FILL HER BOSOM WITH DARK SUSPICIONS CALL TO MIND THEY SAID
2 THE KNOCKER CAUGHT HIS EYE I SHALL LOVE IT AS LONG AS I LIVE CRIED SCROOGE PATTING IT WITH HIS HAND I SCARCELY EVER LOOKED AT IT BEFORE WHAT AN HONEST EXPRESSION IT HAS IN ITS FACE
3 BUT AS HE PLODDED ONWARD THE ATMOSPHERE OF TOWN HAD ITS EFFECT HIS EYE BEGAN TO TAKE NOTE OF THE SIZE OF THE SHOPS GLITTERING UNDER THEIR FALSE FRONTS THE NEW HOUSES BEHIND ROWS OF STIFF YOUNG TREES THE NUMBER AND MAKE OF AUTOMOBILES
4 THUS MY FRIENDS WERE SET FREE FROM ALL THEIR FEARS AND SAW NO MORE OF THEIR WILD FOES FOR SOME TIME ONE DAY A WHIM TOOK THE THREE BAD MEN THAT THEY WOULD GO TO THE MAIN LAND FROM WHENCE THE WILD MEN CAME
5 THEY CAN LEGISLATE LYNCH HUNT NEGROES THROUGH SWAMPS DIVORCE RAILROAD AND RAMPAGE AS MUCH AS EVER THEY CHOOSE THEY DO NOT NEED KNOWLEDGE OF THEIR OWN MILITARY STRENGTH TO BACK THEIR GENIAL LAWLESSNESS THAT REGULAR ARMY
6 MINE CHANGES ALL THE TIME THIS WEEK IT'S ONCE UPON A TIME IN
7 SEPTEMBER EIGHTEEN NINETEEN THE FAMOUS POET JOHN KEATS ARRIVED IN WINCHESTER FOR A SHORT STAY AND WAS SO INSPIRED BY THE CITY'S BEAUTY THAT HE PENNED ONE OF HIS MOST FAMOUS POEMS ODE TO

Each line represents a single sample. The first column is an integer, representing the uttid(just for compatibility with the fixed data format of ESPnet. We don't actually need the uttid information), and the rest of the line is an utterance.

I used the default settings of [conf/tuning/train_lm_transformer2](https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_lm_transformer2.yaml) except that the `batch_bins` is much smaller

The vocabulary size is 32, only containing a\~z and 6 other symbols. 

I print out the inputs for Transformer in the following way
```python

X_lens = (X != self.pad_value).sum(dim=-1)
print(X_Lens.cpu().detach().numpy().tolist())
# examples of output
#     [100, 100, 100]
#     [112, 112, 112]
```
It means that all sentences in current minibatch are in the same length. This could happen but is not the general case because the lengths of sentences in the same minibatch could vary. I suspect that ESPnet chunk the sentences in the same minibatch into the same length but I haven't seen the codes to do it, and also not sure about whether this chunk operation is widely adopted in language modeling or is just used in ESPnet.

One thing to notice is that, before using ESPnet, I used my own codes to train language models on the same corpus, since language models are really easy to implement. I encountered the same problem of loss plateatuing. The loss plateaued after 2 or 3 epoches, with a minimum value of about 1.1\~1.3. I'm not sure the value is already low enough or not for language modelling, but I think it's still quiet high according to my experience.

Any help is appreciated. Thanks! : )

**Comment by xiabingquan** [comment]
> Did you tune the learning rate etc.? Also, I suggest you (virtually) increase the batch size through `accum_grad ` https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_lm_transformer2.yaml#L17

I didn't change anything except for the `batch_bins`. Okay, I'll shut down the `accum_grad` first and see how things work. Thanks for your reply.

**Comment by xiabingquan** [comment]
> Did you tune the learning rate etc.? Also, I suggest you (virtually) increase the batch size through `accum_grad ` https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_lm_transformer2.yaml#L17

I turned off the `accum_grad` and changed learning rate from `1e-3` to `1e-4`, but it still doesn't seem to be working. Here's the loss from tensorboard
![image](https://user-images.githubusercontent.com/53991699/178551965-2e2b9a54-52ff-495e-aedc-54f742759bb7.png)
The cross entropy loss drops from 3.64 to 2.3 in the first 1k iterations. It's just too fast. I'm not sure whether the loss will be NaN again. The training is still in the 1st epoch.

**Comment by sw005320** [comment]
> I turned off the `accum_grad`

`accum_grad` is to increase the batch size practically to make your trial similar to the large GPU memory trial. You said that your batch bins are smaller than that in the config file. Then, you have to increase `accum_grad`.

**Comment by xiabingquan** [comment]
> > I turned off the `accum_grad`
> 
> `accum_grad` is to increase the batch size practically to make your trial similar to the large GPU memory trial. You said that your batch bins are smaller than that in the config file. Then, you have to increase `accum_grad`.

ok. Trying it now.

**Comment by xiabingquan** [comment]
I got another question. Will the learning rate scales with the number of devices(which is a quite common operation)? For example, the learning rate in the configuration file is 1e-4, and we have 8 devices. Is the final learning rate still 1e-4 or 8e-4? I guess it's still 1e-4 becuase I didn't find any part of the codes scales learning rates.
